### PR TITLE
GH-106008: Make implicit boolean conversions explicit

### PR DESCRIPTION
**Pull request by brandtbucher**
...and specialize them!

This adds a new `TO_BOOL` bytecode that prefixes all `UNARY_NOT`, `POP_JUMP_IF_TRUE`, and `POP_JUMP_IF_FALSE` instructions, which now require an exact boolean. We also use a spare bit in `COMPARE_OP`'s oparg to indicate whether the result should be converted to bool (this saves a `TO_BOOL` for most branches, and is a no-op for all `COMPARE_OP` specializations).

["0% faster"](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20230621-3.13.0a0-f9fa498/bm-20230621-linux-x86_64-brandtbucher-to_bool-3.13.0a0-f9fa498-vs-base.png). Stats show a [93.5% hit rate](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20230621-3.13.0a0-f9fa498/bm-20230621-azure-x86_64-brandtbucher-to_bool-3.13.0a0-f9fa498-pystats.md) for the new instructions.

----
:books: Documentation preview :books:: https://cpython-previews--106003.org.readthedocs.build/

* Issue: gh-106008

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -284,15 +284,93 @@ dummy_func(
...
             int err = PyObject_IsTrue(value);
             DECREF_INPUTS();
             ERROR_IF(err < 0, error);
-            if (err == 0) {
-                res = Py_True;
+            res = err ? Py_True : Py_False;
+        }
+
+        inst(TO_BOOL_BOOL, (unused/1, unused/2, value -- value)) {
+            // Coolest (and dumbest-named) specialization ever:
```
Comment (markshannon) [inline Python/bytecodes.c]: True, but the not the most useful comment for someone trying to understand the code 🙂

File: Python/bytecodes.c [hunk 2]
```diff
@@ -284,15 +284,93 @@ dummy_func(
...
+        inst(TO_BOOL_LIST, (unused/1, unused/2, value -- res)) {
+            DEOPT_IF(!PyList_CheckExact(value), TO_BOOL);
+            STAT_INC(TO_BOOL, hit);
+            res = Py_SIZE(value) ? Py_True : Py_False;
+            DECREF_INPUTS();
+        }
+
+        inst(TO_BOOL_NONE, (unused/1, unused/2, value -- res)) {
+            // This one is a bit weird, because we expect *some* failures...
+            // it might be worth combining with TO_BOOL_ALWAYS_TRUE somehow:
```
Comment (markshannon) [inline Python/bytecodes.c]: I think we decided it wasn't, as it reflects the underlying type instability when doing `if x:` as a stand in for `if x is None`.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -284,15 +284,93 @@ dummy_func(
...
+            DEOPT_IF(!Py_IsNone(value), TO_BOOL);
+            STAT_INC(TO_BOOL, hit);
+            res = Py_False;
+        }
+
+        inst(TO_BOOL_STR, (unused/1, unused/2, value -- res)) {
+            DEOPT_IF(!PyUnicode_CheckExact(value), TO_BOOL);
+            STAT_INC(TO_BOOL, hit);
+            if (Py_Is(value, &_Py_STR(empty))) {
```
Comment (markshannon) [inline Python/bytecodes.c]: Use `value == &_Py_STR(empty)`. The semantics is `value == ""`, not `value is ""`. 
In general I wouldn't use `Py_Is` except when you want the exact semantics of Python's `x is y`.

File: Python/bytecodes.c [hunk 4]
```diff
@@ -284,15 +284,93 @@ dummy_func(
...
+            else {
+                assert(Py_SIZE(value));
+                DECREF_INPUTS();
+                res = Py_True;
+            }
+        }
+
+        inst(TO_BOOL_ALWAYS_TRUE, (unused/1, version/2, value -- res)) {
+            // This one is a bit weird, because we expect *some* failures...
+            // it might be worth combining with TO_BOOL_NONE somehow:
```
Comment (markshannon) [inline Python/bytecodes.c]: See comment above.

File: Python/bytecodes.c [hunk 5]
```diff
@@ -2188,35 +2275,13 @@ dummy_func(
...
-                DECREF_INPUTS();
-                if (err == 0) {
-                    JUMPBY(oparg);
-                }
-                else {
-                    ERROR_IF(err < 0, error);
-                }
-            }
+            assert(PyBool_Check(cond));
+            JUMPBY(oparg * Py_IsFalse(cond));
```
Comment (markshannon) [inline Python/bytecodes.c:2271]: This is so much more pleasing 🙂

File: Python/specialize.c [hunk 6]
```diff
@@ -107,6 +107,8 @@ _Py_GetSpecializationStats(void) {
     err += add_stat_dict(stats, COMPARE_OP, "compare_op");
     err += add_stat_dict(stats, UNPACK_SEQUENCE, "unpack_sequence");
     err += add_stat_dict(stats, FOR_ITER, "for_iter");
+    err += add_stat_dict(stats, TO_BOOL, "to_bool");
+    err += add_stat_dict(stats, SEND, "send");
```
Comment (markshannon) [inline Python/specialize.c:111]: Thanks, I tend to forget about the stats dict .

**Review by markshannon — COMMENTED** [review]
A few minor comments, otherwise LGTM.

**Comment by markshannon** [comment]
Note for possible future PR:

We could, at the cost of two bits in `tp_flags` avoid the version number and combine the ALWAYS_TRUE and NONE specializations.
We need a `ALWAYS_TRUE_OR_FALSE` and a `IS_TRUE` bit.
Classes that don't override `__bool__` or `__len__` and `None` would set the  `ALWAYS_TRUE_OR_FALSE` bit. The `IS_TRUE` bit would be set to 0 for `None`, and to 1 for always true objects.

Rather than check the version number, check the `ALWAYS_TRUE_OR_FALSE`, then `res = (tp_flags & IS_TRUE) ? Py_True : Py_False;`

For abi4, we could add a per-object bit to handle immutable objects like ints and strings.

File: Python/bytecodes.c (hunk 3 above)
Comment (brandtbucher) [reply Python/bytecodes.c]: Hm, I mean, we *are* checking for the identity of the singleton string here. I'll just change it, though.

File: Python/flowgraph.c [hunk 7]
```diff
@@ -1120,6 +1120,36 @@ get_const_value(int opcode, int oparg, PyObject *co_consts)
...
+            Py_DECREF(newconst);
+            return -1;
+        }
+        if (PyList_Append(consts, newconst)) {
+            Py_DECREF(newconst);
+            return -1;
+        }
+    }
+    Py_DECREF(newconst);
+    return index;
```
Comment (carljm) [inline Python/flowgraph.c]: windows compiler warning here, implicit cast from Py_ssize_t to int

File: Python/flowgraph.c [hunk 8]
```diff
@@ -1540,6 +1598,39 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
...
+                    continue;
+                }
+                break;
+            case UNARY_NOT:
+                if (nextop == TO_BOOL) {
+                    INSTR_SET_OP0(inst, NOP);
+                    INSTR_SET_OP0(&bb->b_instr[i + 1], UNARY_NOT);
+                    continue;
+                }
+                if (nextop == UNARY_NOT) {
```
Comment (TeamSpen210) [inline Python/flowgraph.c:1552]: Is this correct? `not not x` needs to coerce to bool, it can't be a no-op.

File: Python/bytecodes.c (hunk 3 above)
Comment (markshannon) [reply Python/bytecodes.c]: Hypothetically we could have more than one `""` object. I don't think that `" ".strip() is ""` is part of the language spec, so `Py_Is` doesn't add any safety, just obfuscation.

We can also potentially have tagged ints, in which case `Py_Is` would need to become a lot more complex, but `value == &_Py_STR(empty)` would remain efficient.

File: Python/flowgraph.c (hunk 8 above)
Comment (markshannon) [reply Python/flowgraph.c:1552]: `UNARY_NOT` does no coercion, so `UNARY_NOT; UNARY_NOT` is a no-op.

**Comment by brandtbucher** [comment]
Merging is currently blocked on #106250.

**Comment by gvanrossum** [comment]
Hey @brandtbucher, I have a question about this PR. In https://github.com/python/cpython/pull/106393 I had to change the code in `POP_JUMP_IF_TRUE`/`FALSE` from
```
JUMPBY(oparg * Py_IsFalse(cond));
```
to
```
if (Py_IsFalse(cond)) {
    JUMP_POP_DISPATCH(oparg, 1);  // Macro that wraps JUMPBY()
}
```
The reason is that the uop executor currently exits whenever it jumps, and your original code from this PR *always* jumps.

Did you (or @markshannon) have a reason to prefer the `oparg * Py_IsFalse(cond)` version over the conditional?

If there's no deep reason I'll keep it the way I coded it up; but if there is (maybe it came out faster in a micro-benchmark?) then I suppose I could fix it another way in the uop interpreter (e.g. only exiting if the jump offset is nonzero).

**Comment by markshannon** [comment]
That the implementation of branches is itself branchless has a certain aesthetic appeal 🙂. TBH, that's the main reason.

The multiplication form will be quicker for unpredictable branches, and slower for predictable ones in the tier 1 interpreter.
I doubt it makes any difference in terms of overall performance, and will likely be irrelevant once the tier 2 interpreter does most of the work.

**Pull request merged**